File: src/infinicore/ops/argmax/argmax.cc
#include "infinicore/ops/argmax.hpp"

#include "../../utils.hpp"

#include <stdexcept>

namespace infinicore::op {

common::OpDispatcher<ArgMax::schema> &ArgMax::dispatcher() {
    static common::OpDispatcher<ArgMax::schema> dispatcher_;
    return dispatcher_;
}

void ArgMax::execute(Tensor out, Tensor input, long long dim) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(out, input);

    // 归一化 dim（支持负数）
    Size ndim = input->ndim();
    if (ndim == 0) {
        throw std::runtime_error("ArgMax does not support scalar tensors.");
    }

    long long d = dim;
    if (d < 0) {
        d += static_cast<long long>(ndim);
    }
    if (d < 0 || d >= static_cast<long long>(ndim)) {
        throw std::runtime_error("ArgMax dim is out of range.");
    }

    // 仅支持输入为 F32 / I32，输出为 I64
    auto in_dtype = input->dtype();
    if (in_dtype != DataType::F32 && in_dtype != DataType::I32) {
        throw std::runtime_error("ArgMax only supports F32 and I32 input dtypes for now.");
    }
    if (out->dtype() != DataType::I64) {
        throw std::runtime_error("ArgMax output tensor must have dtype=I64.");
    }

    // 保证输入为连续内存，便于 kernel 简化实现
    Tensor input_contiguous =
        input->is_contiguous() ? input : input->contiguous();

    infinicore::context::setDevice(out->device());

    auto device_type = out->device().getType();
    auto fn = dispatcher().lookup(device_type);
    if (fn == nullptr) {
        throw std::runtime_error("No ArgMax implementation found for device type.");
    }

    fn(out, input_contiguous, d);
}

Tensor argmax(Tensor input, long long dim) {
    Size ndim = input->ndim();
    if (ndim == 0) {
        throw std::runtime_error("ArgMax does not support scalar tensors.");
    }

    long long d = dim;
    if (d < 0) {
        d += static_cast<long long>(ndim);
    }
    if (d < 0 || d >= static_cast<long long>(ndim)) {
        throw std::runtime_error("ArgMax dim is out of range.");
    }

    Shape in_shape = input->shape();
    Shape out_shape;
    out_shape.reserve(ndim > 0 ? ndim - 1 : 0);
    for (Size i = 0; i < ndim; ++i) {
        if (static_cast<long long>(i) == d) {
            continue;
        }
        out_shape.push_back(in_shape[i]);
    }

    auto out = Tensor::empty(out_shape, DataType::I64, input->device());
    argmax_(out, input, d);
    return out;
}

void argmax_(Tensor out, Tensor input, long long dim) {
    ArgMax::execute(out, input, dim);
}

} // namespace infinicore::op

File: python/infinicore/ops/argmax.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def argmax(input: Tensor, dim: int | None = None, keepdim: bool = False) -> Tensor:
    """
    ArgMax 前端接口，语义对齐 `torch.argmax` 的常用用法：

    - 当 dim 为 None 时：会在展平后的维度上做 ArgMax，相当于 `torch.argmax(input.view(-1))`；
    - 当 dim 为整数时：沿指定维度做 ArgMax，输出 shape 为去掉该维度后的形状；
    - keepdim=True 时，会在返回结果中重新插入长度为 1 的该维度。
    """
    if dim is None:
        # 全局 ArgMax：先展平成 1D，再在 dim=0 上做归约
        # 注意：必须先 contiguous()，否则对于带步长的张量 view 会报错
        flat = input.contiguous().view([input.numel()])
        out = Tensor(_infinicore.argmax(flat._underlying, 0))

        if keepdim:
            # 保留维度：将结果 reshape 为 [1, 1, ..., 1]
            return out.view([1] * input.ndim)

        return out

    # 规范化 dim，与 PyTorch 行为一致
    ndim = input.ndim
    if dim < 0:
        dim = dim + ndim

    out = Tensor(_infinicore.argmax(input._underlying, dim))

    if keepdim:
        # 在 dim 位置插入一个长度为 1 的维度
        out_shape = list(out.shape)
        out_shape.insert(dim, 1)
        return out.view(out_shape)

    return out


